pages/1_orders_and_revenue.py

```diff
@@ -16,51 +16,180 @@
+## Ensure all date columns are in datetime format
+date_columns = ['created_at', 'customer_created_at', 'payment_at', 'subscription_period_started_at', 'subscription_period_ended_at']
+for col in date_columns:
+    if col in data.columns:
+        data[col] = pd.to_datetime(data[col], errors='coerce')
 
 st.divider()
 
-# Everything below this point will be generated by the assigned analyst. 
-# The framework has been set, but no visualization or data processing has been applied.
-# Please perform any data processing in this file and not within the filter files. We can discuss upon completion if it makes sense to add any code to the filters file.
-
-## KPI Metrics which need to be updated.
+# KPI Metrics
 with st.container():
     col1, col2, col3, col4 = st.columns(4)
 
+    total_revenue = data['total_amount'].sum()
+    number_of_orders = data['header_id'].nunique()
+    number_of_customers = data['customer_id'].nunique()
+    min_created_at = data['created_at'].min()
+    max_created_at = data['created_at'].max()
+    new_customers = data[(data['customer_created_at'] >= min_created_at) & (data['customer_created_at'] <= max_created_at)].shape[0]
+
     with col1:
-        st.metric(label="**Total Revenue**", value="TBD")
+        st.metric(label="**Total Revenue**", value=f"${total_revenue:,.2f}")
 
     with col2:
-        st.metric(label="**Number of Orders**", value="TBD")
+        st.metric(label="**Number of Orders**", value=number_of_orders)
 
     with col3:
-        st.metric(label="**Number of Customers**", value="TBD")
+        st.metric(label="**Number of Customers in Period**", value=number_of_customers)
 
     with col4:
-        st.metric(label="**New Customers**", value="TBD")
+        st.metric(label="**New Customers**", value=new_customers)
+
+st.divider()
+
+# Time series charts
+data['month'] = data['created_at'].dt.to_period('M').dt.to_timestamp()
 
-# Time series charts need to be built out
 with st.container():
     row1_col1, row1_col2 = st.columns(2)
     row2_col1, row2_col2 = st.columns(2)
 
     with row1_col1:
         st.markdown("**Total Revenue Over Time**")
+        revenue_over_time = data.groupby('month')['total_amount'].sum().reset_index()
+        fig = px.bar(revenue_over_time, x='month', y='total_amount')
+        fig.update_yaxes(tickprefix='$', tickformat='~s', title_text='Revenue', range=[0, revenue_over_time['total_amount'].max() * 1.1])
+        fig.update_xaxes(title_text='Month', tickformat='%b %Y')
+        st.plotly_chart(fig, use_container_width=True)
 
     with row1_col2:
         st.markdown("**Total Orders Over Time**")
+        orders_over_time = data.groupby('month')['header_id'].nunique().reset_index()
+        fig = px.bar(orders_over_time, x='month', y='header_id')
+        fig.update_yaxes(title_text='Orders', range=[0, orders_over_time['header_id'].max() * 1.1])
+        fig.update_xaxes(title_text='Month', tickformat='%b %Y')
+        st.plotly_chart(fig, use_container_width=True)
```

Comment:
Thanks for making these bar charts as I believe that tells a more comprehensive story. However, these charts now look very similar and leads to making the charts seem not that exciting. 

What are your thoughts on making these combo charts and adding a line to each of these which will show the average revenue and average order by month?

Reply:
Additionally, what are your thoughts on adding permanent labels to make this easier to read right away? Similar to the following:

![image](https://github.com/user-attachments/assets/733bae8e-2044-4332-8d0c-1d0303e05a22)

This is not required, but could be a nice feature to add to the accessibility of the charts. [Here](https://github.com/fivetran-streamlit/standardized_billing_model/blob/fa8af19e0c720792af04c06ca119acc522d2b5fd/pages/billing_report.py#L166-L177) is some code that shows how I added those text labels to the chart.

Reply:
updated!

Reply:
@fivetran-reneeli the data labels look great and make this chart way easier to read at first view. Thanks for that update!

> What are your thoughts on making these combo charts and adding a line to each of these which will show the average revenue and average order by month?

I was thinking we keep the two separate charts (one for revenue and one for orders), but then add a line component to each which highlights the rolling overage for month over month. So the Revenue chart would have a line graph included in it which would show the average revenue per month and how it changes month over month. 

I noticed you took a different approach and combined revenue and orders into one chart. The revenue being the bar chart and the orders the line. I like this approach as it helps to reduce repetitive information as both charts were essentially telling the same story.

Reply:
Oh I see, I interpreted as combining revenue and orders into one to reduce the duplicity of the 2 charts. But glad you like it!